std/haxe/macro/CallbackBuilder.hx

```diff
@@ -0,0 +1,41 @@
+package haxe.macro;
+
+import haxe.macro.Expr;
+
+class CallbackBuilder {
+	#if macro
+	static public function build() {
+		switch (Context.getLocalType()) {
+			case TInst(_, [Context.follow(_) => TFun(args, ret)]):
+				var argsComplexTypes = [for (a in args) {
+					if (a.opt)
+						TOptional(Context.toComplexType(a.t));
+					else
+						Context.toComplexType(a.t);
+				}];
+				var retComplexType = Context.toComplexType(ret);
+				var types = [for (len in 0...args.length) argsComplexTypes.slice(0, len+1)];
+
+				// add Void->X
+				types.unshift([macro:Void]);
```

Comment:
Shouldn't this be an empty array? It's the old difference between `Void -> Void` (a function with no arguments) and `(Void) -> Void` (a function with one argument of type `Void`).

I can never remember how we deal with that exactly, but in general what you want is a `TFunction([], ret)`.